### PR TITLE
Proposing structured field overrider API

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**

/kind api-change

**What this PR does / why we need it**:
This PR proposes the API change to address the requirement of #5449.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
For example, if you want to modify the database's host and port in the following ConfigMap:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: example-config
data:
  db-config.yaml: |
    database:
      host: localhost
      port: 3306
```

You can do that by leveraging the proposing API:
```yaml
apiVersion: karmada.io/v1alpha1
kind: OverridePolicy
metadata:
  name: example-configmap-override
spec:
  resourceSelectors:
    - apiVersion: v1
      kind: ConfigMap
      name: example-config
  overrideRules:
    - targetCluster: {}
      overriders:
        fieldOverrider:
          - fieldPath: /data/db-config.yaml
            yaml:
              - subPath: /database/host
                operator: replace
                value: "remote-db.example.com"
              - subPath: /database/port
                operator: replace
                value: "3307"
```
The OverridePolicy declares the replace the /database/host with `remote-db.example.com` and port with `3307`.

The expected result should be:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: example-config
data:
  db-config.yaml: |
    database:
      host: remote-db.example.com
      port: 3307
```

**Does this PR introduce a user-facing change?**:

```release-note
API Change: Introduced `FieldOverrider` to both OverridePolicy and ClusterOverridePolicy, which provides the ability to override structured data nested in manifest like ConfigMap or Secret.
```

### CODE REVIEW

**Comment by sophiefeifeifeiya** [comment]
The API looks good. btw, `overriders` needs to be added.


```
apiVersion: karmada.io/v1alpha1
kind: OverridePolicy
metadata:
  name: example-configmap-override
spec:
  resourceSelectors:
    - apiVersion: v1
      kind: ConfigMap
      name: example-config
  overrideRules:
    - overriders:
        fieldOverrider:
          - fieldPath: /data/db-config.yaml
            yaml:
              - subPath: /database/host
                operator: replace
                value: "remote-db.example.com"
              - subPath: /database/port
                operator: replace
                value: "3307"
```

**Comment by RainbowMango** [comment]
Thanks. Updated.

File: pkg/apis/policy/v1alpha1/override_types.go [hunk 1]
```diff
@@ -255,6 +263,64 @@ const (
...
+type FieldOverrider struct {
+	// FieldPath specifies the initial location in the instance document where the operation should take place.
+	// The path uses RFC 6901 for navigating into nested structures. For example, the path "/data/db-config.yaml"
+	// specifies the configuration data key named "db-config.yaml" in a ConfigMap: "/data/db-config.yaml".
+	// +required
+	FieldPath string `json:"fieldPath"`
+
+	// JSON represents the operations performed on the JSON document specified by the FieldPath.
+	// +optional
+	JSON []JSONPatchOperation `json:"json,omitempty"`
```
[the remark this answers is not in the document]
Comment (sophiefeifeifeiya) [reply pkg/apis/policy/v1alpha1/override_types.go:279]: Just choose one, as fieldPath points to a unique data format.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/override_types.go:280]: Also, add a comment to clarify this.

**Comment by RainbowMango** [comment]
Given our decision on #5449, we can now move this forward.

**Pull request merged**